timemachine/fe/aligned_potential.py

```diff
@@ -0,0 +1,398 @@
+        interpolate.linear_interpolation(w0, w1, jnp.interp(lamb, x, lambdas)),
+        interpolate.linear_interpolation(w1, w0, jnp.interp(1.0 - lamb, x, lambdas)),
+    )
+
+
+batch_interpolate_harmonic_bond_params = jax.jit(
+    jax.vmap(interpolate_harmonic_bond_params, in_axes=(0, 0, None, None, 0, 0))
+)
+batch_interpolate_harmonic_angle_params = jax.jit(
+    jax.vmap(interpolate_harmonic_angle_params, in_axes=(0, 0, None, None, 0, 0))
+)
+batch_interpolate_periodic_torsion_params = jax.jit(
+    jax.vmap(interpolate_periodic_torsion_params, in_axes=(0, 0, None, 0, 0))
+)
+batch_interpolate_chiral_atom_params = jax.jit(
+    jax.vmap(interpolate_chiral_volume_params, in_axes=(0, 0, None, None, 0, 0))
+)
+
+
+@jax.jit
+def batch_interpolate_nonbonded_pair_list_params(
+    cutoff,
+    src_params,
+    dst_params,
+    lamb: float,
+):
+    """
+    Interpolate nonbonded pairlists between two systems. This function sets w coordinates to cutoff
+    for dummy atoms, and linearly interpolates charge, sigma, and epsilon parameters.
+
+    Parameters
+    ----------
+    cutoff: float
+        nonbonded cutoff
+
+    src_params: NDArray
+        Array of P x 4-tuples (q,s,e,w) for the source system
+
+    dst_params: NDArray
+        Array of P x 4-tuples (q,s,e,w) for the destination system
+
+    lamb: float
+        scalar between 0 and 1
+
+    Returns
+    -------
+    NDArray
+        P x 4 array of interpolated pairs.
+
+    """
+    src_qlj, src_w = src_params[:, : NBParamIdx.W_IDX], src_params[:, NBParamIdx.W_IDX]
+    dst_qlj, dst_w = dst_params[:, : NBParamIdx.W_IDX], dst_params[:, NBParamIdx.W_IDX]
+
+    is_excluded_src = jnp.all(src_qlj == 0.0, axis=1, keepdims=True)
+    is_excluded_dst = jnp.all(dst_qlj == 0.0, axis=1, keepdims=True)
+
+    # parameters for pairs that do not interact in the src state
+    w = interpolate_w_coord(cutoff, dst_w, lamb)
+    pair_params_excluded_src = jnp.concatenate((dst_qlj, w[:, None]), axis=1)
+
+    # parameters for pairs that do not interact in the dst state
+    w = interpolate_w_coord(src_w, cutoff, lamb)
+    pair_params_excluded_dst = jnp.concatenate((src_qlj, w[:, None]), axis=1)
+
+    # parameters for pairs that interact in both src and dst states
+    w = jax.vmap(interpolate.linear_interpolation, (0, 0, None))(src_w, dst_w, lamb)
+    qlj = interpolate.linear_interpolation(src_qlj, dst_qlj, lamb)
+    pair_params_not_excluded = jnp.concatenate((qlj, w[:, None]), axis=1)
+
+    pair_params = jnp.where(
+        is_excluded_src,
+        pair_params_excluded_src,
+        jnp.where(
+            is_excluded_dst,
+            pair_params_excluded_dst,
+            pair_params_not_excluded,
+        ),
+    )
+
+    return pair_params
+
+
+@dataclass(kw_only=True)
+class AlignedPotential:
+    src_params: Union[NDArray[np.float64], jax.Array]
+    dst_params: Union[NDArray[np.float64], jax.Array]
+    mins: Optional[NDArray[np.float64]]
+    maxes: Optional[NDArray[np.float64]]
```

Comment:
nit: I find the naming a bit unclear; better as `lambda_min`, `lambda_max`?

Reply:
nit: worth adding a `__post_init__` assertion that `np.all(0.0 < mins)`, `np.all(mins < maxes)`, `np.all(maxes < 1.0)`, etc.?

Reply:
A bit torn on preciseness here:

```
        if self.lambda_min is not None and self.lambda_max is not None:
            assert np.all(0.0 <= self.lambda_min)
            assert np.all(self.lambda_min <= self.lambda_max) # not sure if this should be < or <=
            assert np.all(self.lambda_max <= 1.0) 
        else:
            assert self.lambda_min is None and self.lambda_max is None
```

Reply:
> A bit torn on preciseness (of None values, and < vs <=) here:

(edited my comment above; agreed that we need to allow [0, 1] inclusive)

I guess we might allow `min == max` too, in the case where we want an instantaneous change in the parameter?

Reply:
Okay, I think we're aligned then (i.e. python code above meets both our expectations).